### ISSUE TITLE
kdb global-mount: mounts everything to postcommit

### ISSUE THREAD

**Issue by mpranj** [ElektraInitiative/libelektra#885]
I tried mounting the `dbus` plugin globally. It has the placements: `infos/placements = postgetstorage postcommit`.

I expected that `kdb global-mount dbus` would mount the dbus plugin to both global positions.

When I run `kdb global-mount dbus` it mounts dbus only to the `postcommit ` placement, but not to `postgetstorage`. (when I say mount I mean `kdb export system/elektra/globalplugin` shows some configuration for it)

`kdb global-mount` seems to always write to the same key in the configuration (inside the `postcommit` placement). The other placements are never touched, even though tracer has lots of them (`pregetstorage postgetstorage presetstorage precommit postcommit prerollback postrollback`).

- Elektra Version: master

Output:
[kdb-export.txt](https://github.com/ElektraInitiative/libelektra/files/418956/kdb-export.txt)

**Comment by markus2330** [comment]
Thank you for reporting!

The second (unrelated) issue you reported: Yes, `kdb global-mount` currently overwrites what was done before. Please split the issues. (And remove it from here, so that the threads are easy to read.)

**Comment by tom-wa** [comment]
just looked at it, it seem like global-mount always overrides the last plugin
but dbus gets correctly mounted to posgetstorage and postcommit

```
kdb global-mount dbus
kdb export system/elektra/globalplugins 

....

postcommit/user/plugins =
postcommit/user/plugins/#0 = spec
postcommit/user/plugins/#0/placements = spec
postcommit/user/plugins/#0/placements/error =
postcommit/user/plugins/#0/placements/get = postgetstorage
postcommit/user/plugins/#0/placements/set = presetstorage
postcommit/user/plugins/#1 = dbus
postcommit/user/plugins/#1/placements = dbus
postcommit/user/plugins/#1/placements/error =
postcommit/user/plugins/#1/placements/get = postgetstorage
postcommit/user/plugins/#1/placements/set = postcommit

....

kdb global-mount tracer
kdb export system/elektra/globalplugins

....

postcommit/user/plugins =
postcommit/user/plugins/#0 = spec
postcommit/user/plugins/#0/placements = spec
postcommit/user/plugins/#0/placements/error =
postcommit/user/plugins/#0/placements/get = postgetstorage
postcommit/user/plugins/#0/placements/set = presetstorage
postcommit/user/plugins/#1 = tracer
postcommit/user/plugins/#1/placements = tracer
postcommit/user/plugins/#1/placements/error = prerollback postrollback
postcommit/user/plugins/#1/placements/get = pregetstorage postgetstorage
postcommit/user/plugins/#1/placements/set = postcommit precommit presetstorage

...
```

**Comment by tom-wa** [comment]
>kdb global-mount seems to always write to the same key in the configuration (inside the postcommit placement). The other placements are never touched, even though tracer has lots of them (pregetstorage postgetstorage presetstorage precommit postcommit prerollback postrollback).

thats actually the list plugins plugin configuration (that one that you get when you add plugin configurations without the `-c` flag) , no the real mountpoints

**Comment by tom-wa** [comment]
Ok, it's an issue with the global-mount tools description.
global-mount always overrides everything, is if you want to mount dbus + tracer you need to run:
```
kdb global-mount dbus tracer
kdb export system/elektra/globalplugins

....

postcommit/user/plugins =
postcommit/user/plugins/#0 = spec
postcommit/user/plugins/#0/placements = spec
postcommit/user/plugins/#0/placements/error =
postcommit/user/plugins/#0/placements/get = postgetstorage
postcommit/user/plugins/#0/placements/set = presetstorage
postcommit/user/plugins/#1 = dbus
postcommit/user/plugins/#1/placements = dbus
postcommit/user/plugins/#1/placements/error =
postcommit/user/plugins/#1/placements/get = postgetstorage
postcommit/user/plugins/#1/placements/set = postcommit
postcommit/user/plugins/#2 = tracer
postcommit/user/plugins/#2/placements = tracer
postcommit/user/plugins/#2/placements/error = prerollback postrollback
postcommit/user/plugins/#2/placements/get = pregetstorage postgetstorage
postcommit/user/plugins/#2/placements/set = precommit postcommit presetstorage

...
```

**Comment by mpranj** [comment]
Ok so it does work that way.

Nevertheless: on the left-hand side it says **postcommit**. I'm probably not familiar with the way that the list plugin works, but that's quite confusing.

**Comment by tom-wa** [comment]
it's not really just the list plugin, it works for all plugins in a similar way
the the `list` plugin runs `elektraPluginGetConfig` it behaves exactly like all other plugins and returns the `user/plugins/...` keyset that is below postcommit. kinda similar to:
```
kdb mount /tmp/test.ini system/test.ini ini section=always,meta=
kdb export system/elektra/mountpoints/system\\/test.ini/getplugins

"#0#resolver" =
"#5#ini#ini#" =
"#5#ini#ini#/config" =
"#5#ini#ini#/config/meta" =
"#5#ini#ini#/config/section" = always

kdb export system/elektra/mountpoints/system\\/test.ini/setplugins

"#0#resolver" =
"#5#ini" =
"#6#sync#sync#" =
"#7#resolver" =
```

**Comment by markus2330** [comment]
Please try to separate the issues: please discuss overwriting issues in #886

About the `user` config below postcommit:I do not find where this is described in `src/plugins/list/README.md`, so I think there is a documentation issue here.

**Comment by mpranj** [comment]
The get/set/error list from the list plugin is one thing.

What confuses me is that `kdb global-mount` writes configuration only to `postcommit`. From the configuration it looks like only the `postcommit` hook is executed.

What actually happens, afaik, is that the configuration is just placed there, but the list plugin is executed on all hooks with that same config. Is this correct?

**Comment by markus2330** [comment]
I think thats easily explainable: The config always is added to the first plugin, because this is the place where the plugin gets instantiated. The other places refer back to the first plugin.

I have to admit, however, that the syntax for the global plugins can be improved: it seems like it is not possible to distinguish if you want to instantiate a new plugin or refer back (you can do this for backend-mountpoints)). For the current setup (at every position there is always the same list plugin instance), this feature, however, is not needed.

**Comment by mpranj** [comment]
Ok thanks for the clarification. I think I can work with this.

Seems like #885 and #886 are docu improvement tasks rather than anything else.

**Comment by markus2330** [comment]
I think its important to factor in what someone working with a tool the first time intuitively expected. In both cases, we should improve something (maybe only in long term), so please describe what you *expected* in non-trivial cases (e.g. how to add and remove multiple plugins, especially when mounting the same plugin multiple times such as `kdb gmount dbus#a dbus#b`).

**Comment by tom-wa** [comment]
>About the user config below postcommit:I do not find where this is described in src/plugins/list/README.md, so I think there is a documentation issue here.

actually it's a global hook thing, not a list thing, but you are right, it's my fault for not writing a proper documentation


about the global-mount tool, i'd say we just need 2 things: add and remove 
+ maybe a edit command
i'll improve the addPlugin/editPlugin functionality in list (after i'm done with the firefox thingy) to make it a little bit more useable

**Comment by markus2330** [comment]
Thank you, sounds good. Yes, please concentrate on "firefox thingy" and fix this and the #684 issue later.

**Comment by markus2330** [comment]
As discussed this could be improved by designing a better configuration interface for the list plugin.
